libcaf_net/caf/net/multiplexer.cpp

```diff
@@ -53,411 +76,614 @@ const short error_mask = POLLRDHUP | POLLERR | POLLHUP | POLLNVAL;
-  auto first = managers_.begin();
-  auto last = managers_.end();
-  auto i = std::find(first, last, mgr);
-  return i == last ? -1 : std::distance(first, i);
-}
+  struct poll_update {
+    short events = 0;
+    socket_manager_ptr mgr;
+  };
 
-ptrdiff_t multiplexer::index_of(socket fd) const noexcept {
-  auto first = pollset_.begin();
-  auto last = pollset_.end();
-  auto i = std::find_if(first, last,
-                        [fd](const pollfd& x) { return x.fd == fd.id; });
-  return i == last ? -1 : std::distance(first, i);
-}
+  using poll_update_map = unordered_flat_map<socket, poll_update>;
 
-middleman& multiplexer::owner() {
-  CAF_ASSERT(owner_ != nullptr);
-  return *owner_;
-}
+  using pollfd_list = std::vector<pollfd>;
 
-actor_system& multiplexer::system() {
-  return owner().system();
-}
+  using manager_list = std::vector<socket_manager_ptr>;
 
-// -- implementation of execution_context --------------------------------------
+  // -- friends ----------------------------------------------------------------
 
-void multiplexer::ref_execution_context() const noexcept {
-  ref();
-}
+  friend class pollset_updater; // Needs access to the `do_*` functions.
 
-void multiplexer::deref_execution_context() const noexcept {
-  deref();
-}
+  // -- constructors, destructors, and assignment operators --------------------
 
-void multiplexer::schedule(action what) {
-  CAF_LOG_TRACE("");
-  if (std::this_thread::get_id() == tid_) {
-    pending_actions_.push_back(what);
-  } else {
-    auto ptr = std::move(what).as_intrusive_ptr().release();
-    write_to_pipe(detail::pollset_updater::code::run_action, ptr);
+  explicit default_multiplexer(middleman* parent) : owner_(parent) {
+    // nop
   }
-}
 
-void multiplexer::watch(disposable what) {
-  watched_.emplace_back(what);
-}
+  // -- implementation of caf::net::multiplexer --------------------------------
+
+  // -- initialization ---------------------------------------------------------
+
+  error init() override {
+    auto pipe_handles = make_pipe();
+    if (!pipe_handles)
+      return std::move(pipe_handles.error());
+    auto updater = pollset_updater::make(pipe_handles->first);
+    auto mgr = socket_manager::make(this, std::move(updater));
+    if (auto err = mgr->start())
+      return err;
+    write_handle_ = pipe_handles->second;
+    pollset_.emplace_back(pollfd{pipe_handles->first.id, input_mask, 0});
+    managers_.emplace_back(mgr);
+    return none;
+  }
 
-// -- thread-safe signaling ----------------------------------------------------
+  // -- properties -------------------------------------------------------------
 
-void multiplexer::start(socket_manager_ptr mgr) {
-  CAF_LOG_TRACE(CAF_ARG2("socket", mgr->handle().id));
-  if (std::this_thread::get_id() == tid_) {
-    do_start(mgr);
-  } else {
-    write_to_pipe(detail::pollset_updater::code::start_manager, mgr.release());
+  /// Returns the number of currently active socket managers.
```

Comment:
Doesn't need the documentation since it's overriding. Same for the other member functions.

```suggestion
```